### ISSUE TITLE
Breaking BC change introduced in qx.ui.layout.Grid with PR #9860 with colspan widths

### ISSUE THREAD

**Issue by level420** [qooxdoo/qooxdoo#9871]
It seems that with PR #9860 layout behaviour of `qx.ui.layout.Grid` changed when having non fixed width widgets like `qx.ui.basic.Atom` within a spanned grid cell.

With the new PR labels seem to grow out of bounds of the spanned columns now.

Here is a playground sample showing the behavior before the PR:
https://tinyurl.com/uqdkx6f

and here is a playground sample after the PR:
https://tinyurl.com/v756dgx

@Holdwyn could you please have a look into the playground samples and see if this is a behaviour which you would expect from your PR. Thank you.

**Comment by Holdwyn** [comment]
@level420 yes, that's what I expect. I can not see that the label grows out of bounds, the label grows with its content and should also expand the grid. That's what happens when the column is not spanned, so why this should not happen with a colspan?

**Comment by level420** [comment]
@Holdwyn but why does the label then need two lines? Shouldn't it be just one line?

**Comment by level420** [comment]
The behavious before the PR as shown in https://tinyurl.com/uqdkx6f also makes sense. This way the spanned widget grows only to the maximum of the spanned columns. 
Would it be possible to make this configurable in some way?

**Comment by Holdwyn** [comment]
> @Holdwyn but why does the label then need two lines? Shouldn't it be just one line?

That's strange and seems more like another bug of the rich behavior. In FF it is one line, in chrome the text is wrapped, so I guess it's not a problem of the "new" behaviour.

**Comment by Holdwyn** [comment]
> The behavious before the PR as shown in https://tinyurl.com/uqdkx6f also makes sense. This way the spanned widget grows only to the maximum of the spanned columns.
> Would it be possible to make this configurable in some way?

I don't think so, set the rich attribute to false, why you would expect that the label stops growing, especially if it does without a colspan?

**Comment by johnspackman** [comment]
@Holdwyn I agree that the behaviour of Label to push the bounds of its parent layout is debatable as to whether it always "make sense" - but what's more important here is that there is backwards compatibility in how the layout works.  Grid layout is one of the most commonly used layouts, and any change in behaviour could have a huge impact.

IMHO the PR should be reversed or a new PR put in place to fix your original issue and maintain BC

**Comment by Holdwyn** [comment]
@johnspackman I think the sample is not the best one to describe the problem or the wrong behavior of the grid layout because it looks ok, but it is not. In the sample no item sets the width, so what I expect, is that all width calculation is done from the inside to the outside. This is perfectly working if you not use colspan. With colspan the REAL width of the label it totaly ignored and the only work around is to set the rich property.
Our app makes immersive use of the grid layout and our only workaround to this problem was to set fixed column widths, that totaly break the capability of qooxdoos localisation and theming capabilities. If someone changes the language or switch the theme all labels with longer text are cut of with ellipsis. It was even worse without fixed column sizes, then we had widgets printed over each other...
What we also noticed is that with rowspans the height was calculated correctly (at least what I understand as correct) what #9860 does is to calculate the width with colspans exactly like the hight with rowspans.
But long story short, if most of you think qooxdoo should have a setting to support, instead of removing a layout bug, it's worth another pull request implementing this setting.

**Comment by level420** [comment]
@Holdwyn we could have a property for `qx.ui.layout.Grid` which controls the BC behaviour like `enlargeSpanWidth` by default set to false, or a per cell option like
```
container.add(widget, { row: 1, column: 2, colSpan: 3, enlargeSpanWidth: true});
```
@Holdwyn I'm not arguing that you PR is false (I'll change the issues title to reflect that). In fact I now think you did the right fix. But it may also be desirable to have the old behavior which in some cases makes sense.

**Comment by level420** [comment]
In order to have the possibility to create a BC version of the grid, I've create PR #9874 which changes some members from private to protected to allow overriding of the changed width calculation method `_fixWidthsColSpan`

**Comment by level420** [comment]
@Holdwyn we've introduced a new property `allowGrowSpannedCellWidth` which, when set to `true`, does what you've implemented in PR #9860. Default is `false` which ensures backwards compatibility.
